chore(e2): rhidp-6404  Add image check and relevant changes action to workflows

gustavolira opened the pull request:
https://issues.redhat.com/browse/RHIDP-6404

## Description

Please explain the changes you made here.

## Which issue(s) does this PR fix

- Fixes #?

## PR acceptance criteria

Please make sure that the following steps are complete:

- [ ] GitHub Actions are completed and successful
- [ ] Unit Tests are updated and passing
- [ ] E2E Tests are updated and passing
- [ ] Documentation is updated if necessary (requirement for new features)
- [ ] Add a screenshot if the change is UX/UI related

## How to test changes / Special notes to the reviewer

nickboldt commented on .github/actions/check-image-and-changes/action.yaml:
```diff
@@ -0,0 +1,60 @@
+      id: sha
+      shell: bash
+      run: |
+        SHORT_SHA=$(git rev-parse --short HEAD)
```
probably better to reuse this action for consistency https://github.com/redhat-developer/rhdh/tree/main/.github/actions/get-sha

nickboldt commented on .github/workflows/pr-build-image.yaml:
```diff
@@ -52,69 +52,41 @@ jobs:
+        if: ${{ steps.check-image.outputs.relevant_changes == 'false' && steps.check-image.outputs.image_exists == 'true' }}
+        run: |
+          echo "Skipping build: Image already exists and no relevant changes detected."
+
```
```suggestion
```

nickboldt commented on .github/workflows/pr.yaml:
```diff
@@ -26,9 +26,47 @@ concurrency:
+        run: yarn run lint:check --continue --affected
+
   build:
     name: Build with Node.js ${{ matrix.node-version }}
```
this is a required test. Can we run the step but have it immediately exit 0 if there's nothing to do?

![image](https://github.com/user-attachments/assets/7e0f733f-d703-4a54-9bf7-00a550565f97)

nickboldt commented on .github/workflows/pr.yaml:
```diff
@@ -61,6 +99,8 @@ jobs:
   test:
     name: Test with Node.js ${{ matrix.node-version }}
```
this is a required test. Can we run the step but have it immediately exit 0 if there's nothing to do?

![image](https://github.com/user-attachments/assets/7e0f733f-d703-4a54-9bf7-00a550565f97)

nickboldt commented on .github/actions/check-image-and-changes/action.yaml:
```diff
@@ -0,0 +1,57 @@
+      id: image-check
+      shell: bash
+      run: |
+        IMAGE_TAG_COMMIT="pr-${{ github.event.number }}-${{ steps.get-sha.outputs.short_sha }}"
```
since the get-sha action commits an env var, you can just use...

```suggestion
        IMAGE_TAG_COMMIT="pr-${{ github.event.number }}-${{ env.SHORT_SHA }}"
```

nickboldt commented on .github/workflows/pr-build-image.yaml:
```diff
@@ -52,69 +57,37 @@ jobs:
           imageName: rhdh-community/rhdh
           imageTags: |
-            type=ref,prefix=pr-,suffix=-${{ env.SHORT_SHA }},event=pr
+            type=ref,prefix=pr-,suffix=-${{ needs.check-image.outputs.short_sha }},event=pr
```
```suggestion
            type=ref,prefix=pr-,suffix=-${{ env.SHORT_SHA }},event=pr
```

nickboldt replied:
since the check-image stage sets `${{ env.SHORT_SHA }}` why not just use that env?

nickboldt commented on .github/workflows/pr-build-image.yaml:
```diff
@@ -52,69 +57,37 @@ jobs:
               owner: context.repo.owner,
               repo: context.repo.repo,
-              body: 'The image is available at:\n* [`quay.io/rhdh-community/rhdh:pr-${{ github.event.number }}`](https://quay.io/rhdh-community/rhdh:pr-${{ github.event.number }}) or\n* [`quay.io/rhdh-community/rhdh:pr-${{ github.event.number }}-${{ env.SHORT_SHA }}`](https://quay.io/rhdh-community/rhdh:pr-${{ github.event.number }}-${{ env.SHORT_SHA }})'
+              body: 'The image is available at:\n* [`quay.io/rhdh-community/rhdh:pr-${{ github.event.number }}`](https://quay.io/rhdh-community/rhdh:pr-${{ github.event.number }}) or\n* [`quay.io/rhdh-community/rhdh:pr-${{ github.event.number }}-${{ needs.check-image.outputs.short_sha }}`](https://quay.io/rhdh-community/rhdh:pr-${{ github.event.number }}-${{ needs.check-image.outputs.short_sha }} ...
```
```suggestion
              body: 'The image is available at:\n* [`quay.io/rhdh-community/rhdh:pr-${{ github.event.number }}`](https://quay.io/rhdh-community/rhdh:pr-${{ github.event.number }}) or\n* [`quay.io/rhdh-community/rhdh:pr-${{ github.event.number }}-${{ env.SHORT_SHA }}`](https://quay.io/rhdh-community/rhdh:pr-${{ github.event.number }}-${{ env.SHORT_SHA }})'
```

nickboldt commented on .github/workflows/pr.yaml:
```diff
@@ -60,30 +98,38 @@ jobs:
     strategy:
       matrix:
-        node-version: [22]
+        node-version: [ 20, 22 ]
```
Why are we re-adding node 20?

```suggestion
        node-version: [ 22 ]
```

nickboldt commented on .github/workflows/pr.yaml:
```diff
@@ -60,30 +98,38 @@ jobs:
+          fi
+
       - name: Setup Node.js
-        uses: actions/setup-node@1d0ff469b7ec7b3cb9d8673fde0c81c44821de2a # v4
```
these changes will just trigger renovate updates. Please revert.

nickboldt commented on .github/workflows/pr.yaml:
```diff
@@ -26,9 +26,47 @@ concurrency:
     name: Build with Node.js ${{ matrix.node-version }}
     runs-on: ubuntu-latest
+    needs: check-image
+    if: ${{ needs.check-image.outputs.relevant_changes == 'true' || needs.check-image.outputs.image_exists == 'false' }}
```
looks like the build step will be completely skipped here. can we do a similar check as you have in pr-build-image.yaml line 61? https://github.com/redhat-developer/rhdh/pull/2575/files#diff-2b0fff9a121c4bf67a04b4a82b44d2f2a5cc48a69d8c299b46072978d7cf1212R61

gustavolira replied:
@nickboldt  You're right! The current logic would completely skip the build step if no relevant changes are detected and the image already exists.

I have updated the workflow to ensure that the job always runs (if: always()) and added an explicit check inside the first step. If the conditions indicate that no build is needed, the job exits early with exit 0, similar to what we did in pr-build-image.yaml at line 61.

This way, the workflow structure remains intact, and we avoid unnecessary builds while still ensuring required jobs are executed properly.

Let me know if you have any concerns! Thanks for the feedback.

gustavolira commented on .github/workflows/pr-build-image.yaml:
```diff
@@ -52,48 +56,14 @@ jobs:
-            echo "relevant_changes=false" >> $GITHUB_ENV
+          if [[ "${{ needs.check-image.outputs.relevant_changes }}" == "false" && "${{ needs.check-image.outputs.image_exists }}" == "true" ]]; then
+            echo "Skipping build: Image already exists and no relevant changes detected."
+            exit 0
```
@nickboldt  The skip-build job is just a notification step and does not control execution flow. The build-image job already has an if: condition that ensures it only runs if relevant changes exist or the image does not exist. So, this workflow will not proceed with an unnecessary build.

nickboldt commented on .github/workflows/pr.yaml:
```diff
@@ -26,9 +26,47 @@ concurrency:
     name: Build with Node.js ${{ matrix.node-version }}
     runs-on: ubuntu-latest
+    needs: check-image
+    if: always()
```
do we need this line if the check-if-skip is on line 79?

gustavolira replied:
@nickboldt The if: always() condition isn't needed since we already have a step inside the job that checks if it should be skipped. I'll remove always() and rely on the existing logic to avoid unnecessary executions. Thanks for pointing it out!

nickboldt commented on .github/workflows/pr-build-image.yaml:
```diff
@@ -27,8 +13,27 @@ env:
   build-image:
     name: Build Image
+    needs: check-image
+    if: ${{ needs.check-image.outputs.relevant_changes == 'true' || needs.check-image.outputs.image_exists == 'false' }}
```
this is a required test, so the same "must run and if skipped should exit 0" rule applies here too. 

![image](https://github.com/user-attachments/assets/0caed329-39b9-4e07-8ec9-01a502c92988)

nickboldt commented on .github/workflows/pr.yaml:
```diff
@@ -26,9 +26,47 @@ concurrency:
+        uses: ./.github/actions/check-image-and-changes
+
+  linting:
+    name: Run Linting & Prettier
```
```suggestion
    name: Checkout, install deps, run Linting & Prettier
```

nickboldt commented on .github/workflows/pr-build-image.yaml:
```diff
@@ -1,17 +1,3 @@
-# Copyright Red Hat, Inc.
```
bit weird to remove a copyright statement... but I guess if you have a reason for this...?

nickboldt reviewed (commented):
Only thing I don't like is that we're now doing the same GH checkout twice per action (first to check if we have to build, then to do the build)

:thinking: 

if we moved the call to `.github/actions/check-image-and-changes` out of the `check-image` job and into the `build-image` job, then you might be able to reuse the same checkout for the build, and the step `Check if build should be skipped` would look at  variable `build-image.check-image.outputs.relevant_changes` instead of `needs.check-image.outputs.relevant_changes` ?

nickboldt commented:
Might have a typo somewhere... 

![image](https://github.com/user-attachments/assets/d31eac21-ee5a-43d4-9293-80c46724944f)

https://github.com/redhat-developer/rhdh/actions/runs/13928032057/job/38977666439 

why was the job skipped? why is it incorrectly formatted?

nickboldt commented on .github/workflows/pr.yaml:
```diff
@@ -105,50 +72,43 @@ jobs:
         with:
           cache-prefix: ${{ runner.os }}-v${{ matrix.node-version }}
 
-      - name: Run monorepo tools
```
why did you remove this and the dockerfile regen check? please explain in the issue description -- this is a new change today that was not in the original PR, where you only removed the prettier and linter checks. 

https://github.com/redhat-developer/rhdh/pull/2575/commits/373dac41c690909770704703f9ebfc1b09a30a52#diff-1eb4e5fd5611777d4e597ef299a1cb5ba8050c28a2dabbd4fbc56205d69e5ddaL90

gustavolira replied:
Apologies for the removal, it was unintentional. I got a bit lost during the refactoring while juggling the test day and updating the PR at the same time. In the process, I mistakenly removed the Dockerfile regeneration check. I've now updated the PR and restored everything that was removed. Thanks for pointing it out!

gustavolira commented:
> Only thing I don't like is that we're now doing the same GH checkout twice per action (first to check if we have to build, then to do the build)
> 
> 🤔
> 
> if we moved the call to `.github/actions/check-image-and-changes` out of the `check-image` job and into the `build-image` job, then you might be able to reuse the same checkout for the build, and the step `Check if build should be skipped` would look at variable `build-image.check-image.outputs.relevant_changes` instead of `needs.check-image.outputs.relevant_changes` ?

Thanks for the suggestion! We’ve implemented the improvement by moving the .github/actions/check-image-and-changes call directly into the build-image job. This allows us to reuse the same checkout step, avoiding duplicate operations. Now, the Check if build should be skipped step references build-image.check-image.outputs.relevant_changes instead of needs.check-image.outputs.relevant_changes

nickboldt commented on .github/workflows/pr-build-image.yaml:
```diff
@@ -43,6 +43,17 @@ jobs:
+        id: check-image
+        uses: ./.github/actions/check-image-and-changes
+
+      - name: Check if build should be skipped
```
a further simplication could be to move this check into the action itself.

then the output of the `check-image-and-changes` is not just two booleans, but an exit code of 0 if the job can be terminated right there. 

Of course your logged message would need to be something like

```
            echo "::notice::Skipping job: Image already exists and no relevant changes detected."
```

nickboldt commented:
Did a quick google for how to exit a GH workflow if something fails and it seems we won't get the results we want with an exit 0:

https://stackoverflow.com/questions/60589373/how-to-force-job-to-exit-in-github-actions-step

Instead of exiting 0, we might have to label all the subsequent steps with a conditional check to see if they should run...

```
steps.check.conclusion == 'failure'
```

gustavolira commented:
> Did a quick google for how to exit a GH workflow if something fails and it seems we won't get the results we want with an exit 0:
> 
> https://stackoverflow.com/questions/60589373/how-to-force-job-to-exit-in-github-actions-step
> 
> Instead of exiting 0, we might have to label all the subsequent steps with a conditional check to see if they should run...
> 
> ```
> steps.check.conclusion == 'failure'
> ```

Thanks for the suggestion! We explored that direction, but unfortunately GitHub Actions doesn't currently support terminating a job from within a composite action using `exit 0`—it only ends the current step, not the job. This is also documented in [this StackOverflow thread](https://stackoverflow.com/questions/60589373/how-to-force-job-to-exit-in-github-actions-step) which you sent

To maintain clarity and avoid repeating `if` conditions on every step, we’ve kept the check inside the first step of the job (`Check if build should be skipped`). This gives us the flexibility to skip the entire job cleanly while still keeping the logic centralized in the action.

That said, we're definitely keeping this idea in mind if GitHub Actions evolves to support true early exits!

nickboldt commented on .github/actions/check-image-and-changes/action.yaml:
```diff
@@ -0,0 +1,57 @@
+  steps:
+    - name: Get the last commit short SHA of the PR
+      id: get-sha
+      uses: ./.github/actions/get-sha  # Chamando a action existente
```
```suggestion
      uses: ./.github/actions/get-sha  # Use existing action
```

nickboldt commented:
> To maintain clarity and avoid repeating if conditions on every step, we’ve kept the check inside the first step of the job (Check if build should be skipped). This gives us the flexibility to skip the entire job cleanly while still keeping the logic centralized in the action.

Makes sense, as long as the check returning 0 DOES actually end the job. I guess we can merge this and confirm it does what we want.
